boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+        """Returns the next available line"""
+        pass
+
+    @abstractmethod
+    def readlines(self, sizehint=0):
+        """Returns a list of all lines from the current position forward"""
+        pass
+
+    @abstractmethod
+    def rollover(self):
+        """Roll file-like-object over into a real temporary file"""
+        pass
+
+    @abstractproperty
+    def buffer(self):
+        """Should return a flo instance"""
+        pass
+
+    @abstractproperty
+    def _rolled(self):
+        """Returns whether the file has been rolled to a real file or not"""
+        pass
+
+    def _get_softspace(self):
+        return self.buffer.softspace
+
+    def _set_softspace(self, val):
+        self.buffer.softspace = val
+
+    softspace = property(_get_softspace, _set_softspace)
+
+    @property
+    def _file(self):
+        return self.buffer
+
+    def tell(self):
+        return self._file.tell()
+
+    def seek(self, pos, mode=0):
+        return self.buffer.seek(pos, mode)
+
+    def close(self):
+        return self.buffer.close()
+
+    def flush(self):
+        return self.buffer.flush()
+
+    def isatty(self):
+        return self.buffer.isatty()
+
+    def next(self):
+        return self.readline()
+
+    @property
+    def closed(self):
+        return self.buffer.closed
+
+    @property
+    def pos(self):
+        return self.tell()
+
+    @property
+    def buf(self):
+        return self.getvalue()
+
+    @property
+    def len(self):
+        """Determine the length of the file"""
+        pos = self.tell()
+        if self._rolled:
+            self.seek(0)
+            val = os.fstat(self.fileno()).st_size
+        else:
+            self.seek(0, os.SEEK_END)
+            val = self.tell()
+        self.seek(pos)
+        return val
+
+    def fileno(self):
+        self.rollover()
+        return self.buffer.fileno()
+
+    def truncate(self, size=None):
+        """
+        Custom version of truncate that takes either no arguments (like the
+        real SpooledTemporaryFile) or a single argument that truncates the
+        value to a certain index location.
+        """
+        if size is None:
+            return self.buffer.truncate()
+
+        if size < 0:
+            raise IOError(EINVAL, "Negative size not allowed")
+
+        # Emulate truncation to a particular location
+        pos = self.tell()
+        self.seek(size)
+        self.buffer.truncate(size)
+        if pos < size:
+            self.seek(pos)
+
+    def getvalue(self):
+        """Return the entire files contents"""
+        pos = self.tell()
+        self.seek(0)
+        val = self.read()
+        self.seek(pos)
+        return val
+
+    def __iter__(self):
+        yield self.readline()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self._file.close()
+
+    def __eq__(self, other):
+        if isinstance(other, SpooledIOBase):
+            return (utf8_encode_if_unicode(self.getvalue()) ==
```

Comment:
On Python 2, `u"hi" == b"hi"` because the byte string (`b"hi"`) is implicitly decoded into a `unicode` object using the system's default encoding, which is `ascii`.  That means this will be `False`:

```python
u'u\2022' == b'\xe2\x80\xa2'
```

and produce a warning about how the comparison fails because both arguments could not be converted to `unicode`.

Your implementation would have `SpooledBytesIO(u'\u2022') == SpooledStringIO('\xe2\x80\xa2') return `True`, breaking the illusion of consistency with Python 2's string types.  
Also, comparing a `SpooledBytesIO` object containing zipped data to a `SpooledStringIO` object containing text could raise a `UnicodeDecodeError`, when I'd expect it to return `False`.

And finally, implicit decoding is hard enough to remember in all circumstances that it should be considered an anti-pattern.